browser/ai_chat/ai_chat_tab_helper.cc

```diff
@@ -0,0 +1,299 @@
+       iter != node->UnignoredChildrenEnd(); ++iter) {
+    AddContentNodesToVector(iter.get(), content_nodes);
+  }
+}
+
+void AddTextNodesToVector(const ui::AXNode* node,
+                          std::vector<std::u16string>* strings) {
+  const ui::AXNodeData& node_data = node->data();
+
+  if (node_data.role == ax::mojom::Role::kStaticText) {
+    if (node_data.HasStringAttribute(ax::mojom::StringAttribute::kName)) {
+      strings->emplace_back(
+          node_data.GetString16Attribute(ax::mojom::StringAttribute::kName));
+    }
+    return;
+  }
+
+  for (const auto* child : node->children()) {
+    AddTextNodesToVector(child, strings);
+  }
+}
+}  // namespace
+
+AIChatTabHelper::AIChatTabHelper(content::WebContents* web_contents)
+    : content::WebContentsObserver(web_contents),
+      content::WebContentsUserData<AIChatTabHelper>(*web_contents) {
+  ai_chat_api_ =
+      std::make_unique<AIChatAPI>(web_contents->GetBrowserContext()
+                                      ->GetDefaultStoragePartition()
+                                      ->GetURLLoaderFactoryForBrowserProcess());
+}
+
+AIChatTabHelper::~AIChatTabHelper() = default;
+
+std::string AIChatTabHelper::GetConversationHistoryAsString() {
+  std::string history_text;
+
+  for (ConversationTurn history : chat_history_) {
+    history_text = base::StrCat({history_text,
+                                 history.character_type == CharacterType::HUMAN
+                                     ? ai_chat::kHumanPrompt
+                                     : ai_chat::kAIPrompt,
+                                 history.text, "\n"});
+  }
+
+  return history_text;
+}
+
+const std::vector<ConversationTurn>& AIChatTabHelper::GetConversationHistory() {
+  return chat_history_;
+}
+
+void AIChatTabHelper::AddToConversationHistory(const ConversationTurn& turn) {
+  chat_history_.push_back(turn);
+
+  for (auto& obs : observers_) {
+    obs.OnHistoryUpdate();
+  }
+}
+
+void AIChatTabHelper::SetArticleSummaryString(const std::string& text) {
+  article_summary_ = text;
+}
+
+void AIChatTabHelper::AddObserver(Observer* observer) {
+  observers_.AddObserver(observer);
+}
+
+void AIChatTabHelper::RemoveObserver(Observer* observer) {
+  observers_.RemoveObserver(observer);
+}
+
+void AIChatTabHelper::RequestSummary() {
+  if (!article_summary_.empty()) {
+    VLOG(1) << __func__ << " Article summary is in cache\n";
+
+    AddToConversationHistory({
+        CharacterType::ASSISTANT,
+        ConversationTurnVisibility::VISIBLE,
+        article_summary_,
+    });
+    return;
+  }
+
+  ui::AXTreeID tree_id = web_contents()->GetPrimaryMainFrame()->GetAXTreeID();
+  content::RenderFrameHost* rfh =
+      content::RenderFrameHost::FromAXTreeID(tree_id);
+
+  rfh->GetMainFrame()->RequestAXTreeSnapshot(
+      base::BindOnce(&AIChatTabHelper::OnSnapshotFinished,
+                     base::Unretained(this)),
+      ui::AXMode::kWebContents, false, 5000, {});
```

Comment:
optional: having comment for arguments would help readers to understand this better 

```suggestion
      ui::AXMode::kWebContents, /* exclude_offscreen= */ false, /* max_node= */ 5000, /* timeout= */ {});
```

Reply:
And should we have timeout too, just in case article were very long?

Reply:
A timeout would be suitable, yes. However, It would be nice to test this on real pages or even write tests before adding an arbitrary timeout.